### ISSUE TITLE
p2p: fix Switch.AddPeer signature

### ISSUE THREAD

**Issue by odeke-em** [tendermint/tendermint#667]
A vestige from this PR https://github.com/tendermint/tendermint/pull/653 that changed Peer to an interface rather than a concrete type is that we changed all prior occurances of `*Peer struct` to `Peer interface{}` but unfortunately we left `(*Switch) AddPeer(peer *peer)`

https://github.com/tendermint/tendermint/blob/7f8240dfde4c7acca445e8dbceddb0b2e9a42e7a/p2p/switch.go#L216

 which means that Switch will never be invoked by any external methods yet when setting up say a BlockchainReactor, in order for it to receive blocks transmitted to it by peers, we need to register which peers talk to it with `bcr.Switch.AddPeer(peer)`

**Comment by ebuchman** [comment]
hmmm. do we need to use the switch? can we not just bcr.AddPeer ? 

Regardless, probably right that this should take a `p2p.Peer` interface type, but might have knock on effects. I'll look into it tomorrow.

**Comment by odeke-em** [comment]
Great question actually. In deed, we can invoke bcr.AddPeer, and that's what am using
```go
bcr.Switch = p2p.NewSwitch(cfg.DefaultP2PConfig())
```

I just happened to stumble on it while testing out something with a switch itself, I noticed this because the switch itself implements `AddPeer(peer.Peer)`. The problem with setting that argument to a p2p.Peer is that we might need to either:
1) Add methods to p2p.Peer like:
  - Addr() net.Addr
  - PubKey() crypto.PubKeyEd25519
  - HandshakeTimeout(ourNodeInfo *NodeInfo, timeout time.Duration)
2) Separate out the individual interfaces listed in 1) and those can be optionally implemented as we go through that method.
3) Deprecate Switch <-- I haven't thought much about this but I'll do so later on and evaluate who uses p2p.Switch directly.
